### PR TITLE
Add ParseText tests and fix unmatched TABS handling

### PR DESCRIPTION
**Pull request by Hitcher**
## Description

Add unit test coverage for `CGUITextLayout::ParseText()` and fix unmatched `[TABS]` handling.

The new test suite exercises matched formatting tags, nested tags, unmatched opening/closing tags, malformed nesting, CR/TABS handling, case transforms, deep nesting, and edge-case inputs.

The tests cover parser behaviour related to PR #28280 and PR #28356. While adding coverage, an inconsistency was identified in `[TABS]` handling. This PR updates the parser so malformed `[TABS]` tags are consumed consistently with the other supported formatting tags.

## Motivation and context

Built and executed the `TestGUITextLayoutFilter` test suite.

- `TestGUITextLayoutFilter` tests pass
- All Jenkins builders pass

## How has this been tested?

Built and executed the `TestGUITextLayoutParseText` test suite.

- 52 `TestGUITextLayoutParseText` tests pass
- All Jenkins builders pass

## What is the effect on users?

Improves parser robustness and consistency when processing malformed label formatting tags.

## Types of change

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

## Checklist:

- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [x] I have added tests to cover my change
- [x] All new and existing tests passed

### CODE REVIEW

**Comment by ksooo** [comment]
Sorry, I'm not the right person to review this.

**Comment by Hitcher** [comment]
> Sorry, I'm not the right person to review this.

No problem, just thought it was best to ask you as it covers testing your fix to my previous PR.

File: xbmc/guilib/test/TestGUITextLayout.cpp [hunk 1]
```diff
@@ -0,0 +1,556 @@
...
+
+// ===========================================================================
+// 11. TABS tag
+// ===========================================================================
+
+TEST(TestGUITextLayoutParseText, MatchedTabsTagConsumed)
+{
+  EXPECT_TRUE(PlainTextLacks("[TABS]1[/TABS]label", "[TABS]"));
+  EXPECT_TRUE(PlainTextLacks("[TABS]1[/TABS]label", "[/TABS]"));
+  EXPECT_TRUE(PlainTextHas("[TABS]1[/TABS]label", "label"));
```
Comment (CrystalP) [inline xbmc/guilib/test/TestGUITextLayout.cpp]: A TABS test should confirm the presence of single or multiple tab characters.

File: xbmc/guilib/test/TestGUITextLayout.cpp [hunk 2]
```diff
@@ -0,0 +1,556 @@
...
+// leaving only the plain character content.
+std::string GetPlainText(const std::string& rawLabel)
+{
+  std::string text = rawLabel;
+  CGUITextLayout::Filter(text);
+  return text;
+}
+
+// Returns true when rawLabel filters to exactly expectedPlain.
+bool FiltersTo(const std::string& rawLabel, const std::string& expectedPlain)
```
Comment (CrystalP) [inline xbmc/guilib/test/TestGUITextLayout.cpp]: Counterproductive function that hides the expected value from the unit test output in case of failure.
Please use EXPECT_EQ(expectedPlain, GetPlainText(input)) directly in the unit tests

I'm not sure how to transform PlainTextLacks and PlainTextHas in a way that provides full info in the output for failures.
Although I don't really see the point for those two, equality is all we need and the tests using PlainTextLacks and PlainTextHas can be rewritten in terms of equality instead.
Once that's done, all tests can be transformed into parametrized unit test suite, as is done in many other test files, which are much easier to read, a lot more compact, and easier to maintain.

File: xbmc/guilib/test/TestGUITextLayout.cpp [hunk 3]
```diff
@@ -0,0 +1,556 @@
...
+//               (boldDepth, italicDepth, etc.) for [B], [I], [UPPERCASE],
+//               [LOWERCASE], [CAPITALIZE], [LIGHT] tags.
+//   PR #28356  – added lookahead guard: an opening tag's depth counter is only
+//               incremented when a matching closing tag exists later in the
+//               string; unmatched closing tags at depth 0 are silently ignored.
+//
+// CGUITextLayout::Filter(std::string&) is used as the test entry point.
+// It is public and static, calls ParseText internally, and strips all style
+// bits to return the plain UTF-8 text — exactly what we need to observe
+// whether tags are being recognised and consumed correctly.
```
Comment (CrystalP) [inline xbmc/guilib/test/TestGUITextLayout.cpp]: All those comments (but also throughout the rest of the file) are examples of what garbear called "overgeneration" somewhere in one of your PRs. It's way too much and with the wrong point of view.
Those need to be human read and edited. The same happens in most code AI generates for me, I always have to rewrite the comments.

Also tell the AI we're not in 2024 anymore :-)

File: xbmc/guilib/test/TestGUITextLayout.cpp (hunk 1 above)
Comment (CrystalP) [reply xbmc/guilib/test/TestGUITextLayout.cpp]: edit: comment retracted, I started reading from the bottom of the changes and misunderstood what those unit test did (they do not test the formatting, only the extraction of the remaining plain text, as the helper function name suggests)

File: xbmc/guilib/GUITextLayout.cpp [hunk 4]
```diff
@@ -488,6 +489,7 @@ void CGUITextLayout::ParseText(const std::wstring& text,
     else if (text.compare(pos, 5, L"TABS]") == 0 && on)
     {
       pos += 5;
+      tabsTag = true;
       const size_t end = text.find(L"[/TABS]", pos);
       if (end != std::wstring::npos)
```
Comment (CrystalP) [inline xbmc/guilib/GUITextLayout.cpp]: Would this work instead of your change? and follows the pattern of the rest of the function
```suggestion
      pos += 5;
      if (on)
      {
        const size_t end = text.find(L"[/TABS]", pos);
        if (end != std::wstring::npos)
        {
          std::string t;
          g_charsetConverter.wToUTF8(text.substr(pos), t);
          tabs = atoi(t.c_str());
          pos = end + 7;
        }
      }
```

**Comment by CrystalP** [comment]
What's the bug? It looked to me like [/TABS] without a matching opening [TABS] would be passed through but should not.
[TABS] without a matching closed tab seem filtered already (the pos += 5; consumes the characters)

Are [ characters reserved and must be escaped to be part of the output? I didn't find documentation in the wiki.
If so then the function should maybe have a catch-all to remove all [whatever] text, even for unknown keywords.

File: xbmc/guilib/test/TestGUITextLayout.cpp (hunk 1 above)
Comment (Hitcher) [reply xbmc/guilib/test/TestGUITextLayout.cpp]: Added multiple tabs to the test as well as unmatched tabs.

File: xbmc/guilib/test/TestGUITextLayout.cpp (hunk 2 above)
Comment (Hitcher) [reply xbmc/guilib/test/TestGUITextLayout.cpp]: Replaced all the helper predicates with direct `EXPECT_EQ`/`EXPECT_NE` calls and converted the tests to parametrised suites as you suggested.

File: xbmc/guilib/test/TestGUITextLayout.cpp (hunk 3 above)
Comment (Hitcher) [reply xbmc/guilib/test/TestGUITextLayout.cpp]: Stripped them back to just the two lines at the top explaining why `Filter` is the entry point and why `[COLOR]` is excluded.
Fixed, copyright year is now 2026.

**Comment by Hitcher** [comment]
> What's the bug?

The stray `[/TABS]` case falls through the else if `(text.compare(pos, 5, L"TABS]") == 0 && on)` check because `on` is false, so it's never matched and gets emitted as literal text. I've added a test to cover that and a comment in the test explaining the behaviour difference from the style tags.

File: xbmc/guilib/GUITextLayout.cpp (hunk 4 above)
Comment (Hitcher) [reply xbmc/guilib/GUITextLayout.cpp]: Changed.

**Comment by garbear** [comment]
How is this looking? Ready to go? Please squash into 1 commit.

File: xbmc/guilib/GUITextLayout.cpp (hunk 4 above)
Comment (CrystalP) [reply xbmc/guilib/GUITextLayout.cpp]: Please explain why not verbatim. the tabsTag variable was preserved and is not needed from what I can tell.
tabs != 0 keeps track of the presence of tab characters and is detected already.

File: xbmc/guilib/test/TestGUITextLayout.cpp [hunk 5]
```diff
@@ -1,556 +1,299 @@
...
+
+TEST_P(TestGUITextLayoutMatchedTags, TagStrippedContentPreserved)
+{
+  EXPECT_EQ(GetParam().expected, Filter(GetParam().input));
+}
+
+INSTANTIATE_TEST_SUITE_P(StyleTags,
+                         TestGUITextLayoutMatchedTags,
+                         testing::Values(
+                           // Basic style tags
```
Comment (CrystalP) [inline xbmc/guilib/test/TestGUITextLayout.cpp]: This is progress, thanks, though we don't need multiple sets of tests with multiple INSTANTIATE_TEST_SUITE etc and there are are still quite a few "loose" tests. 
That said, a few of the tests (like the ones with hundreds of tags) are better alone but they're the exception.
Since all tests take the same parameters input and expected, we can have a single class, single tests list, single INSTANTIATE_TEST_SUITE. The separation between test categories can be made with inline comments inside the tests list, same as in other files.
Nowhere in Kodi do we define the list of tests in the INSTANTIATE_TEST_SUITE_P. Much too large for a function call. It's always a static variable. It also allows finetuning of the formatting if needed and reduces the identation.

**Comment by garbear** [comment]
@kodiai please review

File: xbmc/guilib/test/TestGUITextLayout.cpp (hunk 5 above)
Comment (Hitcher) [reply xbmc/guilib/test/TestGUITextLayout.cpp]: New changes ok?

File: xbmc/guilib/test/CMakeLists.txt [hunk 6]
```diff
@@ -1,4 +1,5 @@
 set(SOURCES TestGUIControlFactory.cpp
-            TestGamesGUIInfo.cpp)
+            TestGamesGUIInfo.cpp
+			TestGUITextLayout.cpp)
```
[the remark this answers is not in the document]
Comment (CrystalP) [reply xbmc/guilib/test/CMakeLists.txt]: please address this, spaces are used for tabs.

File: xbmc/guilib/test/TestGUITextLayout.cpp [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+    {"[LIGHT][LIGHT]text[/LIGHT][/LIGHT]", "text"},
+    {"[UPPERCASE][UPPERCASE]hello[/UPPERCASE][/UPPERCASE]", "HELLO"},
+    {"[LOWERCASE][LOWERCASE]HELLO[/LOWERCASE][/LOWERCASE]", "hello"},
+    {"[CAPITALIZE][CAPITALIZE]hello world[/CAPITALIZE][/CAPITALIZE]", "Hello World"},
+
+    // Skin-wrapped addon values: [B]...[B]value[/B]...[/B]
+    {"[B][B]Episode Title[/B][/B]", "Episode Title"},
+    {"[I][I]Some Value[/I][/I]", "Some Value"},
+    {"[LIGHT][LIGHT]Some Value[/LIGHT][/LIGHT]", "Some Value"},
+    {"[UPPERCASE][UPPERCASE]Some Title[/UPPERCASE][/UPPERCASE]", "SOME TITLE"},
```
Comment (CrystalP) [inline xbmc/guilib/test/TestGUITextLayout.cpp]: this is a repetition of the tests immediately above, please remove.

File: xbmc/guilib/test/TestGUITextLayout.cpp [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+TEST_P(TestGUITextLayoutFilterTags, ReturnsExpectedText)
+{
+  EXPECT_EQ(GetParam().expected, Filter(GetParam().input));
+}
+
+static const std::vector<TagTestParam> filterTestCases{
+    // Plain text
+    {"", ""},
+    {"Hello world", "Hello world"},
+    {"price [3.99] per item", "price [3.99] per item"},
```
Comment (CrystalP) [inline xbmc/guilib/test/TestGUITextLayout.cpp]: This happens to work with the current implementation, but seems like a weakness to me and not a case that must not be broken in the future.
How would one get the string "[B]" formatted for example?
I think we need escaping of the [ character and to reject all combinations that are not official formatting commands.
That would be for a follow up PR but for now I don't know if this test has a place here.
@garbear thoughts?

File: xbmc/guilib/test/TestGUITextLayout.cpp [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+};
+
+INSTANTIATE_TEST_SUITE_P(ParseTextTags,
+                         TestGUITextLayoutFilterTags,
+                         testing::ValuesIn(filterTestCases));
+
+TEST(TestGUITextLayoutFilter, SingleOpenBracket)
+{
+  EXPECT_NO_FATAL_FAILURE(Filter("["));
+}
```
Comment (CrystalP) [inline xbmc/guilib/test/TestGUITextLayout.cpp]: Doesn't make sense, there is no subroutine throwing gtest asserts.
A test for a single open or closed bracket could be added to the parametrized tests - though same question as the previous comment for the [3.99] test.

File: xbmc/guilib/GUITextLayout.cpp [hunk 10]
```diff
@@ -485,16 +486,20 @@ void CGUITextLayout::ParseText(const std::wstring& text,
...
-        tabs = atoi(t.c_str());
-        pos = end + 7;
+        const size_t end = text.find(L"[/TABS]", pos);
+        if (end != std::wstring::npos)
+        {
+          std::string t;
+          g_charsetConverter.wToUTF8(text.substr(pos), t);
+          tabs = atoi(t.c_str());
+          pos = end + 7;
+        }
```
[the remark this answers is not in the document]
Comment (CrystalP) [reply xbmc/guilib/GUITextLayout.cpp:500]: true please make the change.

File: xbmc/guilib/GUITextLayout.cpp [hunk 11]
```diff
@@ -403,6 +403,7 @@ void CGUITextLayout::ParseText(const std::wstring& text,
     KODI::UTILS::COLOR::Color newColor = currentColor;
     bool colorTagChange = false;
     bool newLine = false;
+    bool tabsTag = false;
```
Comment (CrystalP) [inline xbmc/guilib/GUITextLayout.cpp]: I'm asking again, why insist on that variable? Please show when it's needed or remove it.

**Comment by CrystalP** [comment]
The tests look a lot better and are much easier to read, review and maintain now, thanks.

File: xbmc/guilib/GUITextLayout.cpp (hunk 11 above)
Comment (Hitcher) [reply xbmc/guilib/GUITextLayout.cpp]: Kept it because it handles unmatched [TABS]: the tag is consumed but no tabs are emitted.

File: xbmc/guilib/test/TestGUITextLayout.cpp (hunk 9 above)
Comment (Hitcher) [reply xbmc/guilib/test/TestGUITextLayout.cpp]: Removed.

File: xbmc/guilib/test/TestGUITextLayout.cpp (hunk 8 above)
Comment (Hitcher) [reply xbmc/guilib/test/TestGUITextLayout.cpp]: Removed `{"price [3.99] per item", "price [3.99] per item"},`.

File: xbmc/guilib/test/TestGUITextLayout.cpp (hunk 7 above)
Comment (Hitcher) [reply xbmc/guilib/test/TestGUITextLayout.cpp]: Swapped tabs for spaces.

Test removed.

File: xbmc/guilib/GUITextLayout.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (Hitcher) [reply xbmc/guilib/GUITextLayout.cpp:500]: Done.

File: xbmc/guilib/test/TestGUITextLayout.cpp [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+    {"[B][I]text[/B][/I]", "text"},
+    {"[B][B]styled[/B][/B] plain", "styled plain"},
+    {"[B][/B]", ""},
+    {"[B][B][/B][/B]", ""},
+    {"[b]text[/b]", "[b]text[/b]"},
+};
+
+INSTANTIATE_TEST_SUITE_P(ParseTextTags,
+                         TestGUITextLayoutFilterTags,
+                         testing::ValuesIn(filterTestCases));
```
Comment (CrystalP) [inline xbmc/guilib/test/TestGUITextLayout.cpp]:
```suggestion
INSTANTIATE_TEST_SUITE_P(TestGUITextLayoutFilter,
                         TestGUITextLayoutFilterTags,
                         testing::ValuesIn(filterTestCases));
```
so that those tests are grouped with the standalone tests, and follow the existing pattern of other test files.

File: xbmc/guilib/GUITextLayout.cpp (hunk 11 above)
Comment (CrystalP) [reply xbmc/guilib/GUITextLayout.cpp]: tabs is set to 0 at the top of the loop and that's enough.
All tests pass without, including those at lines 76-77, the case you mentioned if I understood correctly.

Comment (Hitcher) [reply xbmc/guilib/GUITextLayout.cpp]: Removed `tabsTag`.

File: xbmc/guilib/test/TestGUITextLayout.cpp (hunk 12 above)
Comment (Hitcher) [reply xbmc/guilib/test/TestGUITextLayout.cpp]: Changed the test suite prefix to `TestGUITextLayoutFilter`.

File: xbmc/guilib/GUITextLayout.cpp [hunk 13]
```diff
@@ -485,16 +485,19 @@ void CGUITextLayout::ParseText(const std::wstring& text,
...
-        pos = end + 7;
+        const size_t end = text.find(L"[/TABS]", pos);
+        if (end != std::wstring::npos)
+        {
+          std::string t;
+          g_charsetConverter.wToUTF8(text.substr(pos, end - pos), t);
+          tabs = atoi(t.c_str());
+          pos = end + 7;
+        }
       }
```
Comment (CrystalP) [inline xbmc/guilib/GUITextLayout.cpp:501]: That's weird, the tests passed when I ran them last week but yes there is a problem, sorry.
In the end, a variable like tabsTag is needed or the below suggestion.

```suggestion
      }
      // equivalent of newStyle = of the other formatting tags to have the tag text discarded
      if (!tabs)
        tabs = -1;
    }
```

**Review by CrystalP — APPROVED** [review]
Looks good, sorry for the last detour!
